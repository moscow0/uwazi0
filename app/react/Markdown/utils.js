export const objectPath = (path, object) =>
  path.split('.').reduce((o, key) => {
    if (!o || !key) {
      return o;
    }
    return o.toJS ? o.get(key) : o[key];
  }, object);

export const logError = (err, propValueOf, propLabelOf) => {
  /* eslint-disable no-console */
  console.error('Error on EntityData: ');
  console.error('value-of: ', propValueOf, '; label-of: ', propLabelOf);
  console.error(err);
  /* eslint-enable no-console */
};
