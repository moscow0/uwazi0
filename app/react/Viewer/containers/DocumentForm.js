import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { actions, MetadataForm } from 'app/Metadata';
import { actions as relationshipActions } from 'app/Relationships';
import { saveDocument } from '../actions/documentActions';

function mapStateToProps({ documentViewer, templates, thesauris }) {
  return {
    model: 'documentViewer.sidepanel.metadata',
    isEntity: !documentViewer.sidepanel.file,
    templateId: documentViewer.sidepanel.metadata.template,
    templates,
    thesauris,
  };
}

function mapDispatchToProps(dispatch, ownProps) {
  const { fileID, onEntitySave = () => {} } = ownProps;
  return bindActionCreators(
    {
      changeTemplate: actions.changeTemplate,
      onSubmit: doc => async (disp, state) => {
        const updateDoc = await saveDocument(doc, fileID)(disp, state);
        disp(relationshipActions.reloadRelationships(doc.sharedId));
        onEntitySave(updateDoc);
      },
    },
    dispatch
  );
}
const connected = connect(mapStateToProps, mapDispatchToProps)(MetadataForm);
export { connected as DocumentForm, mapDispatchToProps };
