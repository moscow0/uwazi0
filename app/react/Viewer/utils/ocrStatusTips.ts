import { t } from 'app/I18N';

const ocrStatusTips = {
  noOcr: () => t('System', 'OCR button tip', null, false),
  unsupportedLang: (language: string) => {
    let tip = "The document's language is not supported.";
    if (language === 'other') tip = 'Please select a language for this document';
    return t('System', tip, null, false);
  },
  cantProcess: (time: string) => `${t('System', 'OCR error tip', null, false)}: ${time}`,
  lastUpdated: (time: string) => `${t('System', 'Last updated', null, false)}: ${time}`,
};

export { ocrStatusTips };
