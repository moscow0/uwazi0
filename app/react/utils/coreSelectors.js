import { createSelector } from 'reselect';

export const selectTemplates = createSelector(
  s => s.templates,
  t => t.toJS()
);
