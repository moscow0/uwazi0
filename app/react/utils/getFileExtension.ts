export const getFileExtension = (filename: string) =>
  filename ? filename.substr(filename.lastIndexOf('.') + 1) : '';
