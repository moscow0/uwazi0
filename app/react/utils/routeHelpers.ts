export function validateHomePageRoute(route: string) {
  const queryParams = '(\\?.*)?)';
  const languageMatch = '(/[a-zA-Z]{2,3})?';
  const library = `(library(/map)?(/table)?/?${queryParams}`;
  const entity = '(entity/.+)';
  const page = '(page/.+)';
  const matcher = new RegExp(`${languageMatch}/(${library}|${page}|${entity})$`);
  return route.match(matcher) !== null;
}
