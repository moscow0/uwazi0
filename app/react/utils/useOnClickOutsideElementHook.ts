import { RefObject, useEffect } from 'react';

export function useOnClickOutsideElement<T extends HTMLElement>(
  ref: RefObject<T>,
  cb: (event: MouseEvent) => void
) {
  useEffect(() => {
    const onClickHandler = (event: MouseEvent) => {
      if (!ref.current || ref.current.contains(event.target as HTMLElement)) {
        return;
      }

      cb(event);
    };

    document.addEventListener('click', onClickHandler);
    return () => {
      document.removeEventListener('click', onClickHandler);
    };
  }, [ref, cb]);
}
