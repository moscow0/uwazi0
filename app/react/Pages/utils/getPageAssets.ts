import rison from 'rison-node';
import { get, has, uniq } from 'lodash';
import api from 'app/Search/SearchAPI';
import { markdownDatasets } from 'app/Markdown';
import { RequestParams } from 'app/utils/RequestParams';
import PagesAPI from '../PagesAPI';
import pageItemLists from './pageItemLists';

type Query = { filters: {}; types: string[]; limit?: string };

interface ListsData {
  params: string[];
  content: string;
  options: { limit?: number }[];
  searchs?: any;
}

const buildQuery = (sanitizedParams: string) => {
  const queryDefault = { filters: {}, types: [] };
  if (sanitizedParams) {
    const query = rison.decode(sanitizedParams.replace('?q=', '') || '()');
    if (typeof query !== 'object') {
      return queryDefault;
    }
    return query;
  }
  return queryDefault;
};

const prepareLists = (content: string, requestParams: RequestParams) => {
  const listsData: ListsData = pageItemLists.generate(content);

  listsData.searchs = Promise.all(
    listsData.params.map((params, index) => {
      const sanitizedParams = params ? decodeURI(params) : '';

      const query: Query = buildQuery(sanitizedParams);

      query.limit = listsData.options[index].limit ? String(listsData.options[index].limit) : '6';
      return api.search(requestParams.set(query));
    })
  );

  return listsData;
};

const replaceDynamicProperties = (pageContent?: string, datasets?: any) => {
  if (!pageContent || !datasets || (!datasets.entityData && !datasets.template)) {
    return { content: pageContent, errors: [] };
  }

  const parsableDatasets = {
    entity: datasets.entityData,
    template: datasets.template,
  };

  const errors: string[] = [];
  const content = pageContent.replace(/\$\{((entity.|template.)[^}^\s]*)\}/g, (match, p) => {
    switch (true) {
      case /entity.metadata.\w*$/.test(p):
        return get(parsableDatasets, `${p}[0].value`);

      case /entity.metadata.\w*.(value|displayValue)$/.test(p): {
        const path = p.split('.');
        const pathEnd = path.pop();
        return get(parsableDatasets, `${path.join('.')}[0].${pathEnd}`);
      }

      case /entity.metadata.\w*\[\d+]$/.test(p):
        return get(parsableDatasets, `${p}.value`);

      case has(parsableDatasets, p):
        return get(parsableDatasets, p);

      default:
        errors.push(match);
        return match;
    }
  });
  return { content, errors };
};

const getPageAssets = async (
  requestParams: RequestParams,
  additionalDatasets?: {},
  localDatasets?: {}
) => {
  const page = await PagesAPI.getById(requestParams);

  const { content, errors } = replaceDynamicProperties(page.metadata?.content, localDatasets);
  page.metadata.content = content;

  const listsData = prepareLists(page.metadata.content, requestParams);

  const dataSets = markdownDatasets.fetch(page.metadata.content, requestParams.onlyHeaders(), {
    additionalDatasets,
  });

  const [pageView, searchParams, searchOptions, datasets, listSearchs] = await Promise.all([
    page,
    listsData.params,
    listsData.options,
    dataSets,
    listsData.searchs,
  ]);
  pageView.scriptRendered = false;

  const itemLists = searchParams.map((p, index) => ({
    params: p,
    items: listSearchs[index].rows,
    options: searchOptions[index],
  }));

  const failedExpressions = uniq(errors).join('\n');
  return {
    pageView,
    itemLists,
    datasets: { ...datasets, ...localDatasets },
    ...(failedExpressions.length && {
      errors: `The following expressions are not valid properties:\n ${failedExpressions}`,
    }),
  };
};

export { getPageAssets };
