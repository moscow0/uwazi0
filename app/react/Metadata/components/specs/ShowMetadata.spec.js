import React from 'react';

import { caseTemplate, matterTemplate } from 'app/Timeline/utils/timelineFixedData';
import { shallow } from 'enzyme';

import { FormatMetadata } from 'app/Metadata/containers/FormatMetadata';
import { ShowMetadata } from '../ShowMetadata';

describe('Metadata', () => {
  let props;

  beforeEach(() => {
    props = {
      entity: {},
    };
  });

  it('should render without timeline by default', () => {
    props.entity = {};

    const component = shallow(<ShowMetadata {...props} />);
    expect(component.find('.metadata-timeline-viewer').parent().props().if).toBe(false);
  });

  it('should render templateType when showType', () => {
    props.showType = true;
    const component = shallow(<ShowMetadata {...props} />).find('.item-info');
    expect(component).toMatchSnapshot();
  });

  it('should render title and templateType when and showTitle', () => {
    props.showTitle = true;
    const component = shallow(<ShowMetadata {...props} />).find('.item-info');
    expect(component).toMatchSnapshot();
  });

  it('should render cejil timeline when template is cejils timeline configured templates', () => {
    props.entity = {
      template: caseTemplate,
    };

    let component = shallow(<ShowMetadata {...props} />).find('.metadata-timeline-viewer');
    expect(component).toMatchSnapshot();

    props.entity = {
      template: matterTemplate,
    };

    component = shallow(<ShowMetadata {...props} />).find('.metadata-timeline-viewer');
    expect(component).toMatchSnapshot();
  });

  it('should hide the label if the property is configured to', () => {
    const component = shallow(<ShowMetadata {...props} />);
    const callback = component.find(FormatMetadata).props().renderLabel;
    expect(callback({ noLabel: true }, 'label')).toBeFalsy();
    expect(callback({}, 'label')).toBe('label');
  });
});
