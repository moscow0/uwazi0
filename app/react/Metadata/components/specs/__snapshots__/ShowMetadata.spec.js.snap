// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Metadata should render cejil timeline when template is cejils timeline configured templates 1`] = `
<dl
  className="metadata-timeline-viewer"
>
  <dd>
    <Connect(TimelineViewer)
      entity={
        Object {
          "template": "58b2f3a35d59f31e1345b48a",
        }
      }
    />
  </dd>
</dl>
`;

exports[`Metadata should render cejil timeline when template is cejils timeline configured templates 2`] = `
<dl
  className="metadata-timeline-viewer"
>
  <dd>
    <Connect(TimelineViewer)
      entity={
        Object {
          "template": "58b2f3a35d59f31e1345b4a4",
        }
      }
    />
  </dd>
</dl>
`;

exports[`Metadata should render templateType when showType 1`] = `
<div
  className="item-info"
>
  <Connect(TemplateLabel) />
</div>
`;

exports[`Metadata should render title and templateType when and showTitle 1`] = `
<div
  className="item-info"
>
  <div>
    <Icon
      className="item-icon item-icon-center"
    />
    <h1
      className="item-name"
    >
      <Connect(DocumentLanguage)
        doc={Immutable.Map {}}
      />
    </h1>
  </div>
</div>
`;
